### PR TITLE
Fix #1355: add general-purpose Feature flag system

### PR DESCRIPTION
**Pull request by tatu-at-datastax**
**What this PR does**:

Adds extensible Feature-Flag system to allow both default, system prop setting, and per-request overrides.

Initially just a single feature is added, "tables":

* As YAML Config entry: `stargate.feature.flags.tables`
* As HTTP header: `Feature-Flag-tables` (case-insensitive)

That is: when starting; may specific fixed enabled/disabled via `application.yaml` or system property; latter like:

```
./mvnw quarkus:dev -Dstargate.jsonapi.operations.database-config.local-datacenter=dc1 \
  -Dquarkus.log.console.darken=2  -Dstargate.feature.flags.tables=true
```

but if no setting specified (and ONLY in that case), per-request `curl` header like:

```
curl -i --location --request POST ... --header 'Feature-Flag-tables: true'
```

can be used to enable/disable processing

**Which issue(s) this PR fixes**:
Fixes #1355

**Checklist**
- [x] Changes manually tested
- [ ] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: src/main/resources/application.yaml
Comment (tatu-at-datastax) [inline src/main/resources/application.yaml:30]: `null` meaning "no default", allowing per-request override, whereas `true` would mean enabled and `false` disabled (neither allowing per-request overrides)

Comment (tatu-at-datastax) [reply src/main/resources/application.yaml]: Actually, better choice is to leave it empty; either works.

**Comment by tatu-at-datastax** [comment]
Ready for initial review: need to think about how this could be tested better.

File: CONFIGURATION.md
Comment (tatu-at-datastax) [inline CONFIGURATION.md:74]: Just fixed incorrect class name I noticed.

File: src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/executor/NamespaceCache.java
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/executor/NamespaceCache.java:25]: Had to remove Tables-enabled feature from here because we have to be able enable/disable it on per-request basis. So we will always load regular Table schema find, and fail request at a later point.
Seems more correct overall anyway; checking here would be too low level.

File: src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatureFlag.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatureFlag.java:25]: The command above says "If no configuration specified (config or request), the feature will be Disabled."

consider: we should prob enforce this by default in the enum and remove the enabledByDefault setting. Concern would be that we end up with a mix of values and it gets confusing.

File: src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatureFlag.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatureFlag.java:35]: can we remove "stargate" pls, happy to use "datapi" or nothing. 

also "x" prefixing has been deprecated , are we sure we want to use it ? https://datatracker.ietf.org/doc/html/rfc6648

Consider "Feature-Flag" ?

File: src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatureFlag.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatureFlag.java:34]: consider forcing lcase here, because the feature name is used raw in the header name and other places.

File: src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatures.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatures.java:18]: what is the situation where the fromConfig is null ? could this lead to a subtle bug where the config is missed ?

consider enforcing non-null

nit: also, consider making ctor private if we want to construct via the static factories.

File: src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatures.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatures.java:35]: if I read this correctly, if the feature is disabled in config it is not possible to enable via request. You mention the three ways for controlling features the other day, may be best to spell this out very clearly at the top fo this class.

File: src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/executor/NamespaceCache.java
Comment (amorton) [reply src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/executor/NamespaceCache.java:25]: sounds reasonable

File: src/main/resources/application.yaml
Comment (amorton) [inline src/main/resources/application.yaml:28]: not: can we expand the docs here, or point to expanded docs on the DataApiFeatures class to explain the logic for setting values.

File: src/test/java/io/stargate/sgv2/jsonapi/TestConstants.java
Comment (amorton) [inline src/test/java/io/stargate/sgv2/jsonapi/TestConstants.java:44]: question - how easy / desirable is it to be able to turn on features in a DataApiFeatures class so we can run unit test without needing the things the happen in the DseTestResource class ?

**Review by amorton — CHANGES_REQUESTED** [review]

**Review by amorton — CHANGES_REQUESTED** [review]
generally like where this is going, just comments on names and some details.

File: src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java:21]: Yes, agreed, should change as suggested. Have to think about it a bit ...

File: src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java:77]: Yes, is due to vertx http server impl that does it for HTTP standard reason.

What I am thinking is that we only document use of lower-case header name for usage, and leave ability for case-insensitivity as impl detail: this since we are not exposing features to actual end users. If case-insensitivity wasn't provided I wouldn't bother actually adding it here.

File: src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatureConfig.java
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatureConfig.java:11]: Makes sense, will change.

File: src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatureFlag.java
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatureFlag.java:25]: Problem is that precedence is 3-layered:

1. Explicit value from FeaturesConfig (map keyed by feature)
2. Explicit value from request header if no (1)
3. Default if neither (1) or (2).

I originally thought of just leaving (3) to be `false` for simplicity; I can still do that and drop `enabledByDefault` as it is not strictly needed -- and can be added if/when needed (or something else that does what is needed).

File: src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatureFlag.java
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatureFlag.java:34]: I'll add validation.

File: src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatures.java
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatures.java]: Can/will change to `private`. We get "config" from Quarkus, but I can add `null` check in caller -- but I think it probably better be there just since config mapping might indeed get `null` if our `application.yaml` has empty value (which is reasonable if we do not define any enabled features).

File: src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java:77]: Hmmh. While I'd prefer all lower-case, looks like general best practice seems to be mixed case.

File: src/main/resources/application.yaml
Comment (tatu-at-datastax) [reply src/main/resources/application.yaml]: Fixed link wrt name, but thinking of documenting only/mostly on code side.

File: src/test/java/io/stargate/sgv2/jsonapi/TestConstants.java
Comment (tatu-at-datastax) [reply src/test/java/io/stargate/sgv2/jsonapi/TestConstants.java]: I am thinking that using `DseTestResource` reflects actual usage better, and as long as we do not have tons of testing for alternative setting (that is, most tests can use specific default), works for now.

Otherwise we may end up adding test scaffolding that differs from actual running.
This wrt ITs.

**Comment by tatu-at-datastax** [comment]
Ok: situation at the end of week:

1. Things work wrt Config, Header overrides, with manual testing
2. Addressed doc/comment suggestions
3. Figured out how to do ITs, to verify same as (1)
4. Ran out of time to address one last major thing ("I really want to get the DataApiRequestInfo onto the CommandContext and remove the dependency injection. ") -- hoping to check that on monday.
5. Aside from (4), ready for re-review!

File: src/main/resources/application.yaml
Comment (amorton) [reply src/main/resources/application.yaml]: works for me

File: src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java
Comment (amorton) [reply src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java:21]: ok, we def should not block this PR , it's a nice to have

File: src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java
Comment (amorton) [reply src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java]: I am on the edge of my seat here, seems to be ??? seems to be...

File: src/main/java/io/stargate/sgv2/jsonapi/config/feature/ApiFeature.java
Comment (vkarpov15) [inline src/main/java/io/stargate/sgv2/jsonapi/config/feature/ApiFeature.java:32]: Note to self that this is the header name

**Review by vkarpov15 — COMMENTED** [review]
LGTM :+1:

File: src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatureFlag.java
Comment (amorton) [reply src/main/java/io/stargate/sgv2/jsonapi/config/feature/DataApiFeatureFlag.java]: discussed - drop the enabledByDefault and always default to false. the mental model is these are named boolean values, that default to false.

**Comment by amorton** [comment]
reviewed in person, Done for me . Will approve, and then we need to get the error object V2 flag to use features config

**Comment by tatu-at-datastax** [comment]
Last changes; will merge once CI passes.

File: src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java:21]: Will come back for a look later today but will merge this PR now to unblock other work.

File: src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java]: hmmmh. Not sure why your GUI showed "seems to be" without second line ("mixed case"). Or is that secret line just shown to me? :-o
(I saw it missing when you screen shared)

**Pull request merged**
